### PR TITLE
Fold negation of integer vectors

### PR DESCRIPTION
**Pull request by s-perron**
Fixes https://github.com/microsoft/DirectXShaderCompiler/issues/5247

### CODE REVIEW

File: source/opt/const_folding_rules.cpp [hunk 1]
```diff
@@ -88,6 +88,40 @@ const analysis::Constant* NegateFPConst(const analysis::Type* result_type,
...
+// Returns a constants with the value |-val| of the given type.  Only works for
+// 32-bit and 64-bit signed integer types.  Returns |nullptr| if an error
+// occurs.
+const analysis::Constant* NegateIntConst(const analysis::Type* result_type,
+                                         const analysis::Constant* val,
+                                         analysis::ConstantManager* const_mgr) {
+  const analysis::Integer* int_type = result_type->AsInteger();
+  assert(int_type != nullptr);
+
+  if (!int_type->IsSigned()) {
```
[the remark this answers is not in the document]
Comment (s-perron) [reply source/opt/const_folding_rules.cpp:100]: I looked at the spec and the `OpSNegate` should still work with unsigned values. I'll update.

File: source/opt/const_folding_rules.cpp [hunk 2]
```diff
@@ -89,37 +89,20 @@ const analysis::Constant* NegateFPConst(const analysis::Type* result_type,
...
-    case 64: {
-      int64_t v = val->AsIntConstant()->GetS64();
-      v = (v == INT64_MIN ? v : -v);  // Avoid undefined behaviour.
-      return const_mgr->GetSInt64Const(-v);
-    }
-    default:
-      return nullptr;
-  }
-  return nullptr;
+  uint64_t new_value = -val->GetSignExtendedValue();
```
Comment (Keenuts) [inline source/opt/const_folding_rules.cpp:103]: implicit conversion of signedness, might want to user a static_cast.

File: source/opt/const_folding_rules.cpp [hunk 3]
```diff
@@ -89,37 +89,20 @@ const analysis::Constant* NegateFPConst(const analysis::Type* result_type,
 }
 
 // Returns a constants with the value |-val| of the given type.  Only works for
-// 32-bit and 64-bit signed integer types.  Returns |nullptr| if an error
-// occurs.
+// 32-bit and 64-bit signed integer types.
```
Comment (Keenuts) [inline source/opt/const_folding_rules.cpp:92]: this is not true anymore right?

**Review by Keenuts — DISMISSED** [review]
Some nits and comment wording, letting you handle that 😊

File: source/opt/constants.h [hunk 4]
```diff
@@ -659,11 +659,11 @@ class ConstantManager {
...
   uint32_t GetSIntConstId(int32_t val);
 
-  // Returns a 32-bit signed integer constant with value |val|.
-  const Constant* GetSIntConst(int32_t val);
-
-  // Returns a 64-bit signed integer constant with value |val|.
-  const Constant* GetSInt64Const(int64_t val);
+  // Returns an integer constant with `bitWidth` and value |val|. If `isSigned`
+  // is true, the constant will be a signed integer. Otherwise it will be
+  // unsigned. If `bitWidth` is less than 32, then on the `bitWidth` lower order
```
[the remark this answers is not in the document]
Comment (s-perron) [reply source/opt/constants.h:664]: I fixed it up. The bitWidth lower order bits of |val| will be used.

**Review by Keenuts — APPROVED** [review]

**Pull request merged**